# Importing libraries

In [1]:
import pandas as pd 
import numpy as np 
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split
from sklearn.model_selection import KFold
from sklearn.model_selection import cross_val_score
from sklearn.model_selection import GridSearchCV
from sklearn.linear_model import LogisticRegression
from sklearn.tree import DecisionTreeClassifier
from sklearn.neighbors import KNeighborsClassifier
from sklearn.svm import SVC
from sklearn.naive_bayes import GaussianNB
from sklearn.pipeline import Pipeline
from sklearn.ensemble import AdaBoostClassifier, GradientBoostingClassifier, RandomForestClassifier, ExtraTreesClassifier
from sklearn.metrics import mean_squared_error
from sklearn.preprocessing import OneHotEncoder, LabelEncoder
from sklearn.discriminant_analysis import LinearDiscriminantAnalysis

In [2]:
data=pd.read_csv('new_car_data.csv')
data

policy_id  subscription_length  vehicle_age  customer_age region_code  \
0       POL045360                  9.3          1.2            41          C8   
1       POL016745                  8.2          1.8            35          C2   
2       POL007194                  9.5          0.2            44          C8   
3       POL018146                  5.2          0.4            44         C10   
4       POL049011                 10.1          1.0            56         C13   
...           ...                  ...          ...           ...         ...   
109683  POL044783                 11.6          0.2            54          C5   
109684  POL044451                 11.3          1.0            61          C5   
109685  POL040484                  7.7          0.0            49         C11   
109686  POL040274                 11.2          1.0            44          C8   
109687  POL057561                  4.0          0.0            51         C10   

        region_density segment model fuel_type      max_torque  ...  \
0                 8794      C2    M4    Diesel   250Nm@2750rpm  ...   
1                27003      C1    M9    Diesel   200Nm@1750rpm  ...   
2                 8794      C2    M4    Diesel   250Nm@2750rpm  ...   
3                73430       A    M1       CNG    60Nm@3500rpm  ...   
4                 5410      B2    M5    Diesel   200Nm@3000rpm  ...   
...                ...     ...   ...       ...             ...  ...   
109683           34738      B1    M8       CNG  82.1Nm@3400rpm  ...   
109684           34738      B2    M6    Petrol   113Nm@4400rpm  ...   
109685            6108       A    M1       CNG    60Nm@3500rpm  ...   
109686            8794      C2    M4    Diesel   250Nm@2750rpm  ...   
109687           73430       A    M1       CNG    60Nm@3500rpm  ...   

              model_pop        Pop_max_torque  level_max_power  \
0         Popular_model    Popular_max_torque   High_max_power   
1       Unpopular_model  Unpopular_max_torque    Low_max_power   
2         Popular_model    Popular_max_torque   High_max_power   
3         Popular_model    Popular_max_torque   High_max_power   
4       Unpopular_model  Unpopular_max_torque    Low_max_power   
...                 ...                   ...              ...   
109683  Unpopular_model  Unpopular_max_torque    Low_max_power   
109684    Popular_model    Popular_max_torque   High_max_power   
109685    Popular_model    Popular_max_torque   High_max_power   
109686    Popular_model    Popular_max_torque   High_max_power   
109687    Popular_model    Popular_max_torque   High_max_power   

       Popular_engine_type level_displacement   size_tuning_radius length_cat  \
0         Most_engine_type  High_displacement  large_tuning_radius       long   
1        least_engine_type  High_displacement  small_tuning_radius       long   
2         Most_engine_type  High_displacement  large_tuning_radius       long   
3         Most_engine_type   low_displacement  small_tuning_radius      short   
4        least_engine_type  High_displacement  large_tuning_radius       long   
...                    ...                ...                  ...        ...   
109683   least_engine_type   low_displacement  small_tuning_radius       long   
109684    Most_engine_type  High_displacement  small_tuning_radius       long   
109685    Most_engine_type   low_displacement  small_tuning_radius      short   
109686    Most_engine_type  High_displacement  large_tuning_radius       long   
109687    Most_engine_type   low_displacement  small_tuning_radius      short   

       size_width weight_cat      rating_cat  
0           wider    heavier  average_rating  
1           wider      heavy     good_rating  
2           wider    heavier  average_rating  
3            wide      heavy     Poor_rating  
4           wider      heavy     good_rating  
...           ...        ...             ...  
109683       wide      heavy     Poor_rating  
109684      wider      heavy     Poor_ra

In [3]:
data.columns

Index(['policy_id', 'subscription_length', 'vehicle_age', 'customer_age',
       'region_code', 'region_density', 'segment', 'model', 'fuel_type',
       'max_torque', 'max_power', 'engine_type', 'airbags', 'is_esc',
       'is_adjustable_steering', 'is_tpms', 'is_parking_sensors',
       'is_parking_camera', 'rear_brakes_type', 'displacement', 'cylinder',
       'transmission_type', 'steering_type', 'turning_radius', 'length',
       'width', 'gross_weight', 'is_front_fog_lights', 'is_rear_window_wiper',
       'is_rear_window_washer', 'is_rear_window_defogger', 'is_brake_assist',
       'is_power_door_locks', 'is_central_locking', 'is_power_steering',
       'is_driver_seat_height_adjustable', 'is_day_night_rear_view_mirror',
       'is_ecw', 'is_speed_alert', 'ncap_rating', 'claim_status',
       'density_bins', 'Segment_bins', 'model_pop', 'Pop_max_torque',
       'level_max_power', 'Popular_engine_type', 'level_displacement',
       'size_tuning_radius', 'length_cat', 'size_width'

In [4]:
#droping columns that won't be used for model building
new_data=data.drop(['region_density', 'segment', 'model', 'displacement',
       'max_torque', 'max_power', 'engine_type','length',
       'width', 'gross_weight','turning_radius','ncap_rating'],axis=1)
new_data

policy_id  subscription_length  vehicle_age  customer_age region_code  \
0       POL045360                  9.3          1.2            41          C8   
1       POL016745                  8.2          1.8            35          C2   
2       POL007194                  9.5          0.2            44          C8   
3       POL018146                  5.2          0.4            44         C10   
4       POL049011                 10.1          1.0            56         C13   
...           ...                  ...          ...           ...         ...   
109683  POL044783                 11.6          0.2            54          C5   
109684  POL044451                 11.3          1.0            61          C5   
109685  POL040484                  7.7          0.0            49         C11   
109686  POL040274                 11.2          1.0            44          C8   
109687  POL057561                  4.0          0.0            51         C10   

       fuel_type  airbags is_esc is_adjustable_steering is_tpms  ...  \
0         Diesel        6    Yes                    Yes     Yes  ...   
1         Diesel        2     No                    Yes      No  ...   
2         Diesel        6    Yes                    Yes     Yes  ...   
3            CNG        2     No                     No      No  ...   
4         Diesel        2     No                    Yes      No  ...   
...          ...      ...    ...                    ...     ...  ...   
109683       CNG        2     No                     No      No  ...   
109684    Petrol        2     No                    Yes      No  ...   
109685       CNG        2     No                     No      No  ...   
109686    Diesel        6    Yes                    Yes     Yes  ...   
109687       CNG        2     No                     No      No  ...   

              model_pop        Pop_max_torque level_max_power  \
0         Popular_model    Popular_max_torque  High_max_power   
1       Unpopular_model  Unpopular_max_torque   Low_max_power   
2         Popular_model    Popular_max_torque  High_max_power   
3         Popular_model    Popular_max_torque  High_max_power   
4       Unpopular_model  Unpopular_max_torque   Low_max_power   
...                 ...                   ...             ...   
109683  Unpopular_model  Unpopular_max_torque   Low_max_power   
109684    Popular_model    Popular_max_torque  High_max_power   
109685    Popular_model    Popular_max_torque  High_max_power   
109686    Popular_model    Popular_max_torque  High_max_power   
109687    Popular_model    Popular_max_torque  High_max_power   

        Popular_engine_type level_displacement   size_tuning_radius  \
0          Most_engine_type  High_displacement  large_tuning_radius   
1         least_engine_type  High_displacement  small_tuning_radius   
2          Most_engine_type  High_displacement  large_tuning_radius   
3          Most_engine_type   low_displacement  small_tuning_radius   
4         least_engine_type  High_displacement  large_tuning_radius   
...                     ...                ...                  ...   
109683    least_engine_type   low_displacement  small_tuning_radius   
109684     Most_engine_type  High_displacement  small_tuning_radius   
109685     Most_engine_type   low_displacement  small_tuning_radius   
109686     Most_engine_type  High_displacement  large_tuning_radius   
109687     Most_engine_type   low_displacement  small_tuning_radius   

       length_cat size_width weight_cat      rating_cat  
0            long      wider    heavier  average_rating  
1            long      wider      heavy     good_rating  
2            long      wider    heavier  average_rating  
3           short       wide      heavy     Poor_rating  
4            long      wider      heavy     good_rating  
...           ...        ...        ...             ...  
109683       long       wide      heavy     Poor_rating  
109684       long      wider      heavy     Poor_rating  
109685      sh

Dropped the columns because I had already binned them to classes 

In [5]:
new_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 109688 entries, 0 to 109687
Data columns (total 41 columns):
 #   Column                            Non-Null Count   Dtype  
---  ------                            --------------   -----  
 0   policy_id                         109688 non-null  object 
 1   subscription_length               109688 non-null  float64
 2   vehicle_age                       109688 non-null  float64
 3   customer_age                      109688 non-null  int64  
 4   region_code                       109688 non-null  object 
 5   fuel_type                         109688 non-null  object 
 6   airbags                           109688 non-null  int64  
 7   is_esc                            109688 non-null  object 
 8   is_adjustable_steering            109688 non-null  object 
 9   is_tpms                           109688 non-null  object 
 10  is_parking_sensors                109688 non-null  object 
 11  is_parking_camera                 109688 non-null  o

In [6]:
#Dropping the policy_id and region_code
new_data.drop(['policy_id','region_code'],axis=1,inplace=True)

In [7]:
new_data.apply(lambda x:len(x.unique()))

subscription_length                 140
vehicle_age                          49
customer_age                         41
fuel_type                             3
airbags                               3
is_esc                                2
is_adjustable_steering                2
is_tpms                               2
is_parking_sensors                    2
is_parking_camera                     2
rear_brakes_type                      2
cylinder                              2
transmission_type                     2
steering_type                         3
is_front_fog_lights                   2
is_rear_window_wiper                  2
is_rear_window_washer                 2
is_rear_window_defogger               2
is_brake_assist                       2
is_power_door_locks                   2
is_central_locking                    2
is_power_steering                     2
is_driver_seat_height_adjustable      2
is_day_night_rear_view_mirror         2
is_ecw                                2


# Encoding

# one hot encoding

In [8]:
column_encode={'fuel_type','airbags','steering_type','density_bins','size_width','rating_cat'}
new_data = pd.get_dummies(new_data, columns=list(column_encode))
new_data.head(3)

subscription_length  vehicle_age  customer_age is_esc  \
0                  9.3          1.2            41    Yes   
1                  8.2          1.8            35     No   
2                  9.5          0.2            44    Yes   

  is_adjustable_steering is_tpms is_parking_sensors is_parking_camera  \
0                    Yes     Yes                Yes               Yes   
1                    Yes      No                Yes               Yes   
2                    Yes     Yes                Yes               Yes   

  rear_brakes_type  cylinder  ... airbags_2 airbags_6 fuel_type_CNG  \
0             Disc         4  ...         0         1             0   
1             Drum         4  ...         1         0             0   
2             Disc         4  ...         0         1             0   

  fuel_type_Diesel fuel_type_Petrol size_width_wide size_width_wider  \
0                1                0               0                1   
1                1                0               0                1   
2                1                0               0                1   

  density_bins_High_density density_bins_Low_density  \
0                         0                        0   
1                         1                        0   
2                         0                        0   

  density_bins_Medium_density  
0                           1  
1                           0  
2                           1  

[3 rows x 50 columns]

# label encoding


In [9]:
le = LabelEncoder()
full_encoded_data =new_data.apply(lambda col: le.fit_transform(col) if col.dtype == 'object' else col)
full_encoded_data

subscription_length  vehicle_age  customer_age  is_esc  \
0                       9.3          1.2            41       1   
1                       8.2          1.8            35       0   
2                       9.5          0.2            44       1   
3                       5.2          0.4            44       0   
4                      10.1          1.0            56       0   
...                     ...          ...           ...     ...   
109683                 11.6          0.2            54       0   
109684                 11.3          1.0            61       0   
109685                  7.7          0.0            49       0   
109686                 11.2          1.0            44       1   
109687                  4.0          0.0            51       0   

        is_adjustable_steering  is_tpms  is_parking_sensors  \
0                            1        1                   1   
1                            1        0                   1   
2                            1        1                   1   
3                            0        0                   1   
4                            1        0                   1   
...                        ...      ...                 ...   
109683                       0        0                   1   
109684                       1        0                   1   
109685                       0        0                   1   
109686                       1        1                   1   
109687                       0        0                   1   

        is_parking_camera  rear_brakes_type  cylinder  ...  airbags_2  \
0                       1                 0         4  ...          0   
1                       1                 1         4  ...          1   
2                       1                 0         4  ...          0   
3                       0                 1         3  ...          1   
4                       0                 1         4  ...          1   
...                   ...               ...       ...  ...        ...   
109683                  0                 1         3  ...          1   
109684                  0                 1         4  ...          1   
109685                  0                 1         3  ...          1   
109686                  1                 0         4  ...          0   
109687                  0                 1         3  ...          1   

        airbags_6  fuel_type_CNG  fuel_type_Diesel  fuel_type_Petrol  \
0               1              0                 1                 0   
1               0              0                 1                 0   
2               1              0                 1                 0   
3               0              1                 0                 0   
4               0              0                 1                 0   
...           ...            ...               ...               ...   
109683          0              1                 0                 0   
109684          0              0                 0                 1   
109685          0              1                 0                 0   
109686          1              0                 1                 0   
109687          0              1                 0                 0   

        size_width_wide  size_width_wider  density_bins_High_density  \
0                     0                 1                          0   
1                     0                 1                          1   
2                     0                 1                          0   
3                     1                 0                          1   
4                     0                 1                          0   
...                 ...               ...                        ...   
109683                1                 0                          1   
109684                0                 1                          1   
109685                1                 0                          0

# Train test and split

In [10]:
X = full_encoded_data.drop('claim_status', axis=1)#features
y = full_encoded_data['claim_status']#target

In [11]:
X_train,X_test,y_train,y_test = train_test_split(X,y,test_size=0.2, random_state=42)
print(X_train.shape)
print(y_train.shape)
print(X_test.shape)
print(y_test.shape)

(87750, 49)
(87750,)
(21938, 49)
(21938,)


# Standardization

In [12]:
# Create a StandardScaler object
scaler = StandardScaler()

# Fit the scaler on the training data
scaler.fit(X_train)

# Perform standardization on the training data
X_train = scaler.transform(X_train)

# Perform standardization on the testing data
X_test = scaler.transform(X_test)

# Model Building

In [13]:
import warnings
warnings.filterwarnings("ignore")


In [14]:
models = [
    ('LR', LogisticRegression()),
    ('LDA', LinearDiscriminantAnalysis()),
    ('KNN', KNeighborsClassifier()),
    ('CART', DecisionTreeClassifier()),
    ('NB', GaussianNB())
]

num_folds = 10
seed = 42
scoring = 'accuracy'
# Variables to store results
results = []
names = []

# Evaluation
for name, model in models:
    kfold = KFold(n_splits=num_folds, shuffle=True, random_state=seed)
    cv_results = cross_val_score(model, X_train, y_train, cv=kfold, scoring=scoring)
    results.append(cv_results)
    names.append(name)
    print(f"{name}: {cv_results.mean():.6f} ({cv_results.std():.6f})")

LR: 0.573835 (0.005574)
LDA: 0.574519 (0.005956)
KNN: 0.883054 (0.002770)
CART: 0.950553 (0.001370)
NB: 0.523214 (0.006777)


# Tuning DecisionTreeClassifier

In [15]:
from sklearn.metrics import classification_report
param_grid = {
    'max_depth': [3, 5, 7, 10, None],
    'min_samples_split': [2, 5, 10, 20],
    'min_samples_leaf': [1, 2, 5, 10],
    'max_features': [None, 'sqrt', 'log2']
}
dtree = DecisionTreeClassifier(random_state=42)
# Create a GridSearchCV instance
grid_search = GridSearchCV(estimator=dtree, param_grid=param_grid, cv=5, scoring='accuracy', n_jobs=-1)

# Fit the GridSearchCV to the data
grid_search.fit(X_train, y_train)
# Evaluate the best model on the test set
best_model = grid_search.best_estimator_
y_pred = best_model.predict(X_test)
print("Classification report for the best model:\n", classification_report(y_test, y_pred))

Classification report for the best model:
               precision    recall  f1-score   support

           0       1.00      0.92      0.96     11084
           1       0.92      1.00      0.96     10854

    accuracy                           0.96     21938
   macro avg       0.96      0.96      0.96     21938
weighted avg       0.96      0.96      0.96     21938



# Trying ensemble models

In [16]:
ensembles = [
    ('AB', AdaBoostClassifier()),
    ('GBM', GradientBoostingClassifier()),
    ('RF', RandomForestClassifier()),
    ('ET', ExtraTreesClassifier())
]

num_folds = 10
seed = 42

scoring = 'accuracy'

results = []
names = []

for name, model in ensembles:
    kfold = KFold(n_splits=num_folds, random_state=seed, shuffle=True)
    cv_results = cross_val_score(model, X_train, y_train, cv=kfold, scoring=scoring)
    results.append(cv_results)
    names.append(name)
    msg = f"{name}: {cv_results.mean():.4f} ({cv_results.std():.4f})"
    print(msg)

AB: 0.6085 (0.0046)
GBM: 0.6279 (0.0069)
RF: 0.9703 (0.0017)
ET: 0.9814 (0.0016)


# final model

In [18]:
from sklearn.metrics import classification_report, accuracy_score


etree = ExtraTreesClassifier(random_state=42)
#fitting the model
etree.fit(X_train, y_train)

# Predicting on the test set
y_pred = etree.predict(X_test)

print("Accuracy score: ", accuracy_score(y_test, y_pred))
print("Classification report:\n", classification_report(y_test, y_pred))


Accuracy score:  0.9826328744643996
Classification report:
               precision    recall  f1-score   support

           0       1.00      0.97      0.98     11084
           1       0.97      1.00      0.98     10854

    accuracy                           0.98     21938
   macro avg       0.98      0.98      0.98     21938
weighted avg       0.98      0.98      0.98     21938



# interpretation of the results:

1.For class 0 (no claim), precision is 1.00, meaning that when the model predicts no claim, it is correct 100% of the time. For class 1 (claim), precision is 0.97, indicating that when the model predicts a claim, it is correct 97% of the time.                      
2.For class 0, recall is 0.97, signifying that the model correctly identifies 97% of all actual no-claim instances. For class 1, recall is 1.00, showing that the model correctly identifies 100% of all actual claim instances.                                           
3.The F1-score for both classes is 0.98, indicating a high balance between precision and recall. It means the model is both accurate and reliable in its predictions across both classes.                  
4.The overall accuracy of the model is 98%, which means that it correctly predicts the claim status for 98% of the cases in the test dataset.                      
5.The macro average for precision, recall and F1-score is 0.98, reflecting the average performance of the model across both classes without considering the imbalance in class distribution. This high value suggests that the model performs well across both classes. The weighted average for precision, recall, and F1-score is also 0.98, taking into account the imbalance in class distribution. It indicates that, on average, the model performs consistently well across the different classes when considering their distribution in the dataset.